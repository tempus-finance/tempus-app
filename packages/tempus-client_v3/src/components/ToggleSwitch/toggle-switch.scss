$toggle-switch-width: 32px;
$toggle-switch-height: 16px;
$toggle-switch-padding: 2px;

$thumb-size: $toggle-switch-height + 2 * $toggle-switch-padding;
$thumb-active-state-x-translation: $toggle-switch-width - $thumb-size;

$background-border-radius: $toggle-switch-height / 2;

.tc__toggle-switch {
  display: flex;
  align-items: center;
}

.tc__toggle-switch__switch-container {
  width: $toggle-switch-width;
  height: $toggle-switch-height;
  margin: $toggle-switch-padding 0;
  display: inline-flex;
  position: relative;

  input[type='checkbox'] {
    width: 100%;
    height: 100%;
    margin: 0;
    top: 0;
    left: 0;
    opacity: 0;
    position: absolute;
    z-index: 1;
  }
}

.tc__toggle-switch__thumb {
  width: $thumb-size;
  height: $thumb-size;
  margin-top: -$toggle-switch-padding;
  position: absolute;
  background: #ffffff;
  border: 1px solid var(--greyLight);
  border-radius: 50%;
  box-sizing: border-box;
  box-shadow: var(--downStrongShadow);
  transition: transform 300ms ease-out;
}

.tc__toggle-switch__thumb-active {
  background: var(--primaryMain);
  transform: translateX($thumb-active-state-x-translation);
  border-color: var(--primaryDark);
}

.tc__toggle-switch__background {
  width: 100%;
  height: 100%;
  background: var(--greyLight);
  border-radius: $background-border-radius;
}

.tc__toggle-switch__background-active {
  background: var(--primaryLight);
}

.tc__toggle-switch__label {
  margin-left: 6px;
}
