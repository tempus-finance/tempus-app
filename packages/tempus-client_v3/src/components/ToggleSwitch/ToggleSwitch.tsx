import { useMemo } from 'react';
import { Typography } from '../shared';
import './toggle-switch.scss';

interface ToggleSwitchProps {
  checked: boolean;
  label?: string;
}

let idCounter = 0;

const ToggleSwitch = (props: ToggleSwitchProps & React.HTMLProps<HTMLInputElement>) => {
  const { checked, label, ...checkboxProps } = props;
  const id = useMemo(() => checkboxProps.id ?? `toggle-switch-${idCounter++}`, [checkboxProps.id]);

  return (
    <span className="tc__toggle-switch">
      <span className="tc__toggle-switch__switch-container">
        <input {...checkboxProps} id={id} checked={checked} type="checkbox" />
        <span className={`tc__toggle-switch__thumb ${checked ? 'tc__toggle-switch__thumb-active' : ''}`}></span>
        <span
          className={`tc__toggle-switch__background ${checked ? 'tc__toggle-switch__background-active' : ''}`}
        ></span>
      </span>
      {label && (
        <label htmlFor={id}>
          <Typography
            variant="body-secondary"
            weight={checked ? 'bold' : 'regular'}
            className="tc__toggle-switch__label"
          >
            {label}
          </Typography>
        </label>
      )}
    </span>
  );
};

export default ToggleSwitch;
