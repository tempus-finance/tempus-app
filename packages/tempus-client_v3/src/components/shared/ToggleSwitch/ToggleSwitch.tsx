import { ChangeEvent, FC, useMemo } from 'react';
import { Typography } from '..';
import './toggle-switch.scss';

export interface ToggleSwitchProps {
  checked: boolean;
  label?: string;
  onChange?: (event: ChangeEvent<HTMLInputElement>) => void;
}

let idCounter = 0;

const ToggleSwitch: FC<ToggleSwitchProps> = props => {
  const { checked, label, onChange } = props;
  const id = useMemo(() => `toggle-switch-${idCounter++}`, []);

  return (
    <span className="tc__toggle-switch">
      <span className="tc__toggle-switch__switch-container">
        <input id={id} checked={checked} type="checkbox" onChange={onChange} />
        <span className={`tc__toggle-switch__thumb ${checked ? 'tc__toggle-switch__thumb-active' : ''}`}></span>
        <span
          className={`tc__toggle-switch__background ${checked ? 'tc__toggle-switch__background-active' : ''}`}
        ></span>
      </span>
      {label && (
        <label htmlFor={id} className="tc__toggle-switch__label">
          <Typography variant="body-secondary" weight={checked ? 'bold' : 'regular'}>
            {label}
          </Typography>
        </label>
      )}
    </span>
  );
};

export default ToggleSwitch;
