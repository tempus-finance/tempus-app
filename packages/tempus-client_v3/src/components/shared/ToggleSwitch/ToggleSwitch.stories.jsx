import React from 'react';

import ToggleSwitch from './ToggleSwitch';

export default {
  title: 'ToggleSwitch',
  component: ToggleSwitch,
  argTypes: {},
};

const style = {
  background: 'rgba(0, 0, 0, 0.1)',
  display: 'flex',
  alignItems: 'center',
  justifyContent: 'center',
  padding: '10px',
};

const Template = args => (
  <div style={style}>
    <ToggleSwitch {...args} />
  </div>
);

export const Unchecked = Template.bind({});
Unchecked.args = {
  checked: false,
  label: 'unchecked',
};

export const Checked = Template.bind({});
Checked.args = {
  checked: true,
  label: 'checked',
};
