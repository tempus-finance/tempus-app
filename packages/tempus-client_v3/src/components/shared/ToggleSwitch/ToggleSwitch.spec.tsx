import { fireEvent, render } from '@testing-library/react';
import ToggleSwitch, { ToggleSwitchProps } from './ToggleSwitch';

const mockOnChange = jest.fn();

const defaultProps: ToggleSwitchProps = {
  checked: false,
  label: 'my label',
  onChange: mockOnChange,
};

const subject = (props: ToggleSwitchProps) => render(<ToggleSwitch {...props} />);

describe('ToggleSwitch', () => {
  it('renders an unchecked toggle switch', () => {
    const { container, queryByText, getByRole } = subject(defaultProps);

    const actualToggleSwitch = container.querySelector('.tc__toggle-switch');
    const actualLabel = queryByText('my label');
    const actualCheckbox = getByRole('checkbox');

    expect(actualToggleSwitch).not.toBeNull();
    expect(actualLabel).not.toBeNull();

    expect(actualCheckbox).not.toBeNull();
    expect(actualCheckbox).not.toBeChecked();

    expect(actualToggleSwitch).toMatchSnapshot();
  });

  it('renders a checked toggle switch', () => {
    const { container, queryByText, getByRole } = subject({ ...defaultProps, checked: true });

    const actualToggleSwitch = container.querySelector('.tc__toggle-switch');
    const actualLabel = queryByText('my label');
    const actualCheckbox = getByRole('checkbox');

    expect(actualToggleSwitch).not.toBeNull();
    expect(actualLabel).not.toBeNull();

    expect(actualCheckbox).not.toBeNull();
    expect(actualCheckbox).toBeChecked();

    expect(actualToggleSwitch).toMatchSnapshot();
  });

  it('calls `onChange` when the toggle switch button is clicked', () => {
    const { getByRole } = subject(defaultProps);

    const actualCheckbox = getByRole('checkbox');
    fireEvent.click(actualCheckbox);

    expect(mockOnChange).toHaveBeenCalledTimes(1);
  });
});
