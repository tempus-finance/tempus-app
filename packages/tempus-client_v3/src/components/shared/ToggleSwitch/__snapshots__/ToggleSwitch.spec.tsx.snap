// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ToggleSwitch renders a checked toggle switch 1`] = `
<span
  class="tc__toggle-switch"
>
  <span
    class="tc__toggle-switch__switch-container"
  >
    <input
      checked=""
      id="toggle-switch-2"
      type="checkbox"
    />
    <span
      class="tc__toggle-switch__thumb tc__toggle-switch__thumb-active"
    />
    <span
      class="tc__toggle-switch__background tc__toggle-switch__background-active"
    />
  </span>
  <label
    class="tc__toggle-switch__label"
    for="toggle-switch-2"
  >
    <div
      style="font-style: normal; font-size: 12px; line-height: 16px; color: rgb(34, 34, 34); font-weight: 700; font-family: 'DM Sans', sans-serif; opacity: 1;"
    >
      my label
    </div>
  </label>
</span>
`;

exports[`ToggleSwitch renders an unchecked toggle switch 1`] = `
<span
  class="tc__toggle-switch"
>
  <span
    class="tc__toggle-switch__switch-container"
  >
    <input
      id="toggle-switch-1"
      type="checkbox"
    />
    <span
      class="tc__toggle-switch__thumb "
    />
    <span
      class="tc__toggle-switch__background "
    />
  </span>
  <label
    class="tc__toggle-switch__label"
    for="toggle-switch-1"
  >
    <div
      style="font-style: normal; font-size: 12px; line-height: 16px; color: rgb(34, 34, 34); font-weight: 400; font-family: 'DM Sans', sans-serif; opacity: 1;"
    >
      my label
    </div>
  </label>
</span>
`;
